# Import Packages

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

# Read Data

In [2]:
# Set working directory
%cd C:/Users/sriha/OneDrive/Desktop/DS/My work/life-expectancy
    
path = "Data/Life Expectancy Data.csv"

df = pd.read_csv(path)

C:\Users\sriha\OneDrive\Desktop\DS\My work\life-expectancy


In [3]:
pd.set_option('max_columns', None)
df.head(5)

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0             62     0.01               71.279624         65.0      1154   
1             64     0.01               73.523582         62.0       492   
2             66     0.01               73.219243         64.0       430   
3             69     0.01               78.184215         67.0      2787   
4             71     0.01                7.097109         68.0      3013   

    BMI   under-five deaths   Polio  Total expenditure  Diphtheria   \
0   19.1                  83    6.0               8.16         65.0   
1   18.6                  86   58.0               8.18         62.0   
2   18.1                  89   62.0               8.13         64.0   
3   17.6                  93   67.0               8.52         67.0   
4   17.2                  97   68.0               7.87         68.0   

    HIV/AIDS         GDP  Population   thinness  1-19 years  \
0        0.1  584.259210  33736494.0                   17.2   
1        0.1  612.696514    327582.0                   17.5   
2        0.1  631.744976  31731688.0                   17.7   
3        0.1  669.959000   3696958.0                   17.9   
4        0.1   63.537231   2978599.0                   18.2   

    thinness 5-9 years  Income composition of resources  Schooling  
0                 17.3                            0.479       10.1  
1                 17.5                            0.476       10.0  
2                 17.7                            0.470        9.9  
3                 18.0                            0.463        9.8  
4                 18.2                            0.454        9.5

In [4]:
# Note that some of the column names have spaces and other anomalies. Keep this in mind while cleaning

df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

## Column Descriptions

**The original dataset contains details of some of the columns. The rest are inferred based on observing the data and similarity to other columns. Here are some column descriptions for the not-so-obvious column names:**

**Life expectancy:** Life expectancy in years for a given country and year

**Adult Mortality:** Number of people dying between the age of 15 and 60, per 1000 population

**Infant deaths:** Number of infants dying per 1000 population

**Alcohol:** Alcohol consumption measured as litres of pure alcohol consumed per capita

**Percentage expenditure:** Expenditure on health as a percentage of GDP

**Hepatitis B:** Hepatitis B percentage immunization coverage among 1-year olds

**Measles:** Number of reported cases of measles per 1000 population

**BMI:** Average BMI for the country's population

**Under-five deaths:** Number of people dying under the age of 5, per 1000 population

**Polio:** Polio percentage immunization coverage among 1-year olds (inferred from similarity to Hep B column)

**Total expenditure:** Expenditure on health as a percentage of total expenditure (note the difference from percentage expenditure column)

**Diptheria:** Diptheria percentage immunization coverage among 1-year olds

**HIV/AIDS:** Unsure. Potentially deaths from HIV/AIDS per 1000 population

**Thinness 1-19 years:** Unsure. Potentially related to BMI somehow, rate of thinness among people aged 1-19 

**Thinness 5-9 years:** Unsure. Potentially rate of thinness among people aged 5-9 years

**Income composition of resources:** Human Development Index in termss of income composition, value ranging from 0 to 1\

**Schooling:** Unsure. Potentially number of years of schooling on average

# Explore Data

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

In [6]:
df.describe()

Year  Life expectancy   Adult Mortality  infant deaths  \
count  2938.000000       2928.000000      2928.000000    2938.000000   
mean   2007.518720         69.224932       164.796448      30.303948   
std       4.613841          9.523867       124.292079     117.926501   
min    2000.000000         36.300000         1.000000       0.000000   
25%    2004.000000         63.100000        74.000000       0.000000   
50%    2008.000000         72.100000       144.000000       3.000000   
75%    2012.000000         75.700000       228.000000      22.000000   
max    2015.000000         89.000000       723.000000    1800.000000   

           Alcohol  percentage expenditure  Hepatitis B       Measles   \
count  2744.000000             2938.000000  2385.000000    2938.000000   
mean      4.602861              738.251295    80.940461    2419.592240   
std       4.052413             1987.914858    25.070016   11467.272489   
min       0.010000                0.000000     1.000000       0.000000   
25%       0.877500                4.685343    77.000000       0.000000   
50%       3.755000               64.912906    92.000000      17.000000   
75%       7.702500              441.534144    97.000000     360.250000   
max      17.870000            19479.911610    99.000000  212183.000000   

              BMI   under-five deaths         Polio  Total expenditure  \
count  2904.000000         2938.000000  2919.000000         2712.00000   
mean     38.321247           42.035739    82.550188            5.93819   
std      20.044034          160.445548    23.428046            2.49832   
min       1.000000            0.000000     3.000000            0.37000   
25%      19.300000            0.000000    78.000000            4.26000   
50%      43.500000            4.000000    93.000000            5.75500   
75%      56.200000           28.000000    97.000000            7.49250   
max      87.300000         2500.000000    99.000000           17.60000   

       Diphtheria      HIV/AIDS            GDP    Population  \
count  2919.000000  2938.000000    2490.000000  2.286000e+03   
mean     82.324084     1.742103    7483.158469  1.275338e+07   
std      23.716912     5.077785   14270.169342  6.101210e+07   
min       2.000000     0.100000       1.681350  3.400000e+01   
25%      78.000000     0.100000     463.935626  1.957932e+05   
50%      93.000000     0.100000    1766.947595  1.386542e+06   
75%      97.000000     0.800000    5910.806335  7.420359e+06   
max      99.000000    50.600000  119172.741800  1.293859e+09   

        thinness  1-19 years   thinness 5-9 years  \
count            2904.000000          2904.000000   
mean                4.839704             4.870317   
std                 4.420195             4.508882   
min                 0.100000             0.100000   
25%                 1.600000             1.500000   
50%                 3.300000             3.300000   
75%                 7.200000             7.200000   
max                27.700000            28.600000   

       Income composition of resources    Schooling  
count                      2771.000000  2775.000000  
mean                          0.627551    11.992793  
std                           0.210904     3.358920  
min                           0.000000     0.000000  
25%                           0.493000    10.100000  
50%                           0.677000    12.300000  
75%                           0.779000    14.300000  
max                           0.948000    20.700000

## Let's try to understand the data from this

There are a lot of inconsistencies and anomalies in this data. We need to understand the data to retain as much useful information, and remove noise. Based on the column descriptions above, we can make some inferences about the data. 

1. **'Infant deaths':** Minimum value of infant deaths cannot be equal to 0. No matter how advanced healthcare in a country is, there will be a non zero value of infant mortality. It is likely that this is missing data in the dataset. 
    - Replace zero values with NaN


2. **'BMI':** Minimum value of 1 and maximum value of 87 is not possible. BMI would very likely range between 15 and 40. Anything below or over is of serious concern. Here, this could likely be faulty data that will have to be dealt with. WHO data from other sources shows countries by avg BMI, and the values range from 20 to 35. 
    - Replace values below 15 or above 40 with NaN


3. **'Under-five deaths':** Similar to infant mortality, the value cannot be equal to 0. 
    - Replace zero values with NaN


4. **'GDP':** Minimum value of 1.68 is unlikely. Further, there seems to be massive variation in the GDP values of the same country, year-to-year. 


5. **'Population':** Minimum value of 34 again is unlikely. Similar to GDP, there is a lot of variation from year to year for the same country which is not likely. 

In [7]:
df['infant deaths'] = df['infant deaths'].replace(0, np.nan)
df[' BMI '] = df.apply(lambda x: np.nan if (x[' BMI '] < 15 or x[' BMI '] > 45) else x[' BMI '], axis=1)
df['under-five deaths '] = df['under-five deaths '].replace(0, np.nan)

In [8]:
# Figure out how much missing data is in the dataset

def find_missing(dataset, return_feature_names=False):
  
    from IPython.display import display

    missing_values = pd.DataFrame(dataset.isnull().sum())
    missing_percent = pd.DataFrame(missing_values/len(dataset)*100)
    missing = pd.concat([missing_values, missing_percent], axis=1)
    missing.columns = ['Missing values', '%']
    missing = missing.sort_values(by='Missing values', ascending=False)

    display(missing[missing['Missing values'] > 0])

    if return_feature_names == True:
        return missing[missing['Missing values'] > 0].index.values.tolist()

In [9]:
find_missing(df)

Missing values          %
 BMI                                       1835  62.457454
infant deaths                               848  28.863172
under-five deaths                           785  26.718856
Population                                  652  22.191967
Hepatitis B                                 553  18.822328
GDP                                         448  15.248468
Total expenditure                           226   7.692308
Alcohol                                     194   6.603131
Income composition of resources             167   5.684139
Schooling                                   163   5.547992
 thinness  1-19 years                        34   1.157250
 thinness 5-9 years                          34   1.157250
Polio                                        19   0.646698
Diphtheria                                   19   0.646698
Adult Mortality                              10   0.340368
Life expectancy                              10   0.340368

# Cleaning Logic

- Overall, the data quality is quite poor. In several cases (example, GDP and population), the standard deviation far exceeds the mean. Population values for a country cannot vary year over year by orders of magnitude.


- In addition to this, several countries have no data for GDP and population (for example, Yemen). 


- With the logic previously defined, BMI has 57% missing values. It is best to drop this column entirely. 


- Life expectancy will be the target variable, hence we cannot impute these values. Rows with null values for this column will be dropped. 


- For the rest of the columns, missing values will be filled by the median value of the column, grouped by country. 


- Note that all the columns that have missing values are of type float. This makes value imputation fairly simple. 

In [10]:
df.drop(' BMI ', inplace=True, axis=1)
df.dropna(subset = ['Life expectancy '], inplace=True)

In [11]:
medians_by_country = df.groupby('Country').median().reset_index()
medians_by_country.head()

Country    Year  Life expectancy   Adult Mortality  \
0          Afghanistan  2007.5             57.80            284.0   
1              Albania  2007.5             75.60             17.5   
2              Algeria  2007.5             73.95            124.5   
3               Angola  2007.5             48.45            367.0   
4  Antigua and Barbuda  2007.5             75.10            143.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0           81.0     0.01               17.029434         64.0    1794.0   
1            1.0     4.95              100.361243         98.0       8.5   
2           20.0     0.45              187.424821         94.5     109.5   
3           85.5     6.35               35.946032         72.0    1322.5   
4            NaN     7.84             1027.575528         99.0       0.0   

   under-five deaths   Polio  Total expenditure  Diphtheria    HIV/AIDS  \
0               111.5   58.0               8.19         62.5        0.1   
1                 1.0   98.0               5.94         98.0        0.1   
2                23.5   94.5               3.84         95.0        0.1   
3               135.5   52.0               3.84         52.5        2.5   
4                 NaN   98.0               4.53         99.0        0.1   

            GDP  Population   thinness  1-19 years   thinness 5-9 years  \
0    321.199783   2924815.0                  18.50                18.30   
1   1390.308084    290456.5                   1.60                 1.65   
2   3031.800100  32440295.0                   6.00                 5.90   
3   1111.730143   2891811.0                   8.55                 8.55   
4  11650.644430         NaN                   3.40                 3.30   

   Income composition of resources  Schooling  
0                           0.4240       8.55  
1                           0.7080      11.80  
2                           0.6935      12.45  
3                           0.4610       7.90  
4                           0.7795      13.90

In [12]:
df.update(df[['Country']].merge(medians_by_country, on='Country',  how='left'), overwrite=False)

find_missing(df)

Missing values          %
infant deaths                               632  21.584699
under-five deaths                           595  20.321038
Population                                  541  18.476776
GDP                                         335  11.441257
Income composition of resources             133   4.542350
Schooling                                   133   4.542350
Hepatitis B                                 120   4.098361
Total expenditure                            22   0.751366
 thinness  1-19 years                        14   0.478142
 thinness 5-9 years                          14   0.478142
Alcohol                                       7   0.239071

While we were able to significantly reduce the number of missing values for some of the columns, we are still unable to remove them all. At this stage, we can repeat the process by filling missing values with medians grouped by year (instead of country). 

Note that to maintain data quality, we should be trying to impute medians by country. For example, the median GDP of a country is likely to be closer to its own 10-year median, vs. the annual median of ALL countries in the world. So imputing by year is a last resort to impute values.

In [13]:
medians_by_year = df.groupby('Year').median().reset_index()
medians_by_year.head()

Year  Life expectancy   Adult Mortality  infant deaths  Alcohol  \
0  2000              71.0            155.0            9.0     3.61   
1  2001              71.2            151.0            9.0     3.81   
2  2002              71.4            146.0            9.0     3.76   
3  2003              71.1            144.0            9.0     4.04   
4  2004              71.2            158.0            9.0     3.81   

   percentage expenditure  Hepatitis B  Measles   under-five deaths   Polio  \
0               35.661251        88.25        47                11.0   88.0   
1               46.830275        88.50        49                10.0   89.0   
2               49.735114        87.00        36                10.0   91.0   
3               55.595439        85.75        44                10.0   91.0   
4               55.361072        88.00        22                10.0   91.0   

   Total expenditure  Diphtheria    HIV/AIDS          GDP  Population  \
0               5.43         86.0        0.1  1164.092895   1265314.5   
1               5.46         89.0        0.1  1417.265820   1352219.0   
2               5.46         89.0        0.1  1148.229882   1391024.0   
3               5.61         91.0        0.1  1286.264450   1458821.0   
4               5.69         89.0        0.1  1360.094916   1239196.0   

    thinness  1-19 years   thinness 5-9 years  \
0                    3.4                  3.4   
1                    3.2                  3.3   
2                    3.4                  3.3   
3                    3.3                  3.3   
4                    3.2                  3.4   

   Income composition of resources  Schooling  
0                            0.606      11.40  
1                            0.631      11.60  
2                            0.634      11.80  
3                            0.640      11.85  
4                            0.649      12.05

In [14]:
df.update(df[['Year']].merge(medians_by_year, on='Year',  how='left'), overwrite=False)

find_missing(df)

Empty DataFrame
Columns: [Missing values, %]
Index: []

We have now eliminated all missing values.

# Create Prepocessor Object

Now that we have a clearly defined logic for our cleaning process, we can creeate a preprocessor object. The purpose of this to create a repeatable method if our model is deployed into production and the data is updated. 

We will start the process from scratch, do a test-train split to avoid data leakage when calculating our median values. 

**Steps:**

1. Read and split data, and separate X and y variables
2. Write a preprocessor object that calculates medians from the train dataset
3. Impute these median values (as discussed above) to the train as well as test sets
4. Save these cleaned datasets for next steps

In [15]:
# Read in dataset
df = pd.read_csv(path)

# Drop rows will null values in target variable
df.dropna(subset=['Life expectancy '], inplace=True)

from sklearn.model_selection import train_test_split

X_raw = df.drop('Life expectancy ', axis=1)
y = df['Life expectancy ']

X_train, X_test, y_train, y_test = train_test_split(X_raw, y, test_size=0.25)

In [16]:
class Preprocessor:

    def __init__(self):
        self.country_medians = pd.DataFrame()
        self.year_medians = pd.DataFrame()


    def fit(self, X, y=None):
        '''
        Learn information from the training data to transform the test data. 
        Only call fit on train data.
        '''

        #Calculate median values by country and year
        self.country_medians = X.groupby('Country').median().reset_index()
        self.year_medians = X.groupby('Year').median().reset_index()

        return self


    def transform(self, X, y=None):
        '''
        Use fit information if applicable and impute values to train and test data.
        '''
        X_new = X.copy()
        X_new.reset_index(inplace=True) # Temporarily reset indices

        # Replace illogical values with NaNs for later imputation
        X_new['infant deaths'] = X_new['infant deaths'].replace(0, np.nan)
        X_new['under-five deaths '] = X_new['under-five deaths '].replace(0, np.nan)

        # Drop columns with >50% missing values
        X_new.drop(' BMI ', inplace=True, axis=1)

        X_new.update(X_new[['Country']].merge(medians_by_country, on='Country',  how='left'), overwrite=False)
        X_new.update(X_new[['Year']].merge(medians_by_year, on='Year',  how='left'), overwrite=False)

        # Restore original indices
        X_new.set_index('index', inplace=True)
        X_new.index.name = None

        return X_new


    def fit_transform(self, X, y=None):
        '''
        Combines fit and transform for use in sklearn pipelines
        '''
        return self.fit(X).transform(X)

In [17]:
prep = Preprocessor()

In [18]:
X_train_clean = prep.fit_transform(X_train)
find_missing(X_train_clean)

Empty DataFrame
Columns: [Missing values, %]
Index: []

In [19]:
X_test_clean = prep.transform(X_test)
find_missing(X_test_clean)

Empty DataFrame
Columns: [Missing values, %]
Index: []

# Save Cleaned Data

In [20]:
# pickle preprocessed train data
X_train_clean.to_pickle('Data/X_train_clean.pkl')

# pickle preprocessed test data
X_test_clean.to_pickle('Data/X_test_clean.pkl')

# pickle train target variable
y_train.to_pickle('Data/y_train.pkl')

# pickle test target variable
y_test.to_pickle('Data/y_test.pkl')